### ISSUE TITLE
UIA handler (backward compatibility): unify public and private modules under a package

### ISSUE THREAD

**Issue by josephsl** [nvaccess/nvda#10916]
Hi,

The following should wait until work on next incompatible version of NVDA begins:

### Is your feature request related to a problem? Please describe.
For years, UIA handler module was split into two parts: a publicly accessible module (UIAHandler.py) and a private module (_UIAHandler.py). This split was necessary in 2011 in order for NVDA to run without UI Automation client support (namely on Windows XP, Server 2003, and old releases of Windows Vista and Server 2008). Additionally, private UIA handler module wasn't used if UIA was forcefully disabled. Lastly, other accessibility API handlers such as JAB and IAccessible are contained inside one module.

### Describe the solution you'd like
Unify public and private UIA handler modules into one file.

### Describe alternatives you've considered
Leave the UIA handler arrangement as is.

### Additional context
Consider that:

1. NVDA supports Windows 7, Server 2008 R2 and later, which comes with a bit more robust UIA support than old Windows releases, but not up to the level as seen with Windows 8/Server 2012 and later.
2. If UIA is forcefully disabled, an error is raised.
3. Other accessibility API handlers are housed inside one file per API, whereas UIA is split in two. This means majority of UIA support improvements are housed inside the private module, with the public module importing everything.

As this proposal deals with an important accessibility API handler, I propose this change as a backwards incompatible change.

Thanks.

**Comment by feerrenrut** [comment]
I understand that the initial rationale for splitting this module no longer applies, but can you add to this the benefits to this refactor? 

As with any refactor there is risk of breakage, or as already pointed out breaking backwards compatibility. The benefits have to outweigh this.

**Comment by josephsl** [comment]
Hi, the biggest benefit to this move is module content consistency with other accessibility API’s: namely editing only one module would be sufficient rather than thinking about public interface versus implementation. Another benefit is code deduplication: debug flag from private UIA interface is imported by the public after all module contents are imported, so this flag is imported twice. Combined, these benefits will allow ease of future maintenance for people wishing to work with UIA support. At the moment there are no known modules that import the private module directly, as NVDA will assume that public module includes everything from the private module.

**Comment by feerrenrut** [comment]
One benefit from having a private module is that we can change the private interface without breaking compatibility. Generally I wouldn't want to refactor code unless it is already being modified for some other purpose.

**Comment by josephsl** [comment]
Hi, true. Note however that in case of UIA handler, since the public module imports everything from the private module, it becomes public (which can be argued for all Python modules as even private variables (apart from mangled ones) can be used and modified). @MIchaelDCurran, any thoughts, since you re the one who split UIA handler?

**Comment by leonardder** [comment]
@josephsl Could you pick this up for 2021.1?

**Comment by josephsl** [comment]
Hi,

The bulk of the upcoming pull request consists of lint fixes.

Thanks.

**Comment by josephsl** [comment]
Hi,

There's one important thing to be aware of: UIA constants in UIA handler come from UI Automation Client COM library (comtypes.gen.UIAutomationClient). I bet many parts of NVDA source code (particularly UIA NVDA objects collection) are using these constants directly without going through UIAHandler.UIA, which is the alias for UIA client library. This issue will become prominent while running linters such as Flake8, as Flake8 will record F405 for up to 80 of these.

I can think of several solutions:

1. Define UIA constants ourselves, similar to how oleacc module defines MSAA roles and states. This will guarantee compatibility and is perhaps the simplest option.
2. Replace UIAHandler.UIA_* and other constants with UIAHandler.UIA.UIA_*, which will involve extensive source code review and may affect several add-ons (possibly one of mine).
3. Define a getattr function inside UIAHandler that will return these constants directly in 2021. In 2022, usage of UIAHandler.UIA_* will be disallowed.

I think it would be best to avoid option 2 for now due to time constraints unless folks can come up with something before 2021.1 beta 1 is released; if we go with option 2, perhaps a deprecation notice should be added in 2021.x. Option 1 may take several hours to a day to implement, verify, and test, and I need to test option 3 with Flake8 to see how it'll respond to UIAHandler.__getattr__.

Thanks.

**Comment by josephsl** [comment]
Hi,

After thinking about Flake8 F405 issue, I think it would be advisable to split this issue into at least two parts:

1. Change NVDA source code to use UIAHandler.UIA.UIA_* constants. I think this should be done first in a 2021.x release.
2. Transfer the private UIA handler module content to the public one, perhaps in 2022.1.

Although defining our own constants would work for now, it isn't really future-proof - UIA interface can change without notice.

Thanks.

**Comment by feerrenrut** [comment]
The comInterfaces constants is actually something that I'm looking at currently. I am planning a change to make the constants discoverable. Code in NVDA should directly use the constants from `comInterfaces/UIAutomationClient.py`.

As for the UIAHandler, I'd like this to be made into a package. Both `UIAHandler.py` and `_UIAHandler.py` should be moved in here. I think parts of `_UIAHandler.py` should be split up. Note there is also a soon to have a PR branch that affects this code, and adds several more UIA related files: `excelWithUIA-rebase`

**Comment by josephsl** [comment]
Hi, in this case, I think it’d be best to revisit this ticket as early as early summer (early winter in southern hemisphere) once the relevant changes are in. Thanks.

**Comment by feerrenrut** [comment]
I think these changes will either have to be in 2021.1 or wait until 2022.1.

**Comment by josephsl** [comment]
Hi, given the complexities discussed above, I vote to push this to 2022.1. Besides, this issue will be impacted by many things, including WinVersion class PR. Thanks.

**Comment by josephsl** [comment]
Hi,

NOw that 2021.1 beta is out, we can revisit this somewhat...

After thinking about this and looking at current state of UIA handler, I think it would be better if we can use package route, the same route taken by IAccessibleHandler package. The new UIAHandler package can include an __init__ to store contents of combined public and private module, along with dedicated files to store UIA constants and such.

Thanks.

**Comment by feerrenrut** [comment]
NV Access doesn't have any plan to work on this, please note, for this to be included in 2022.1 it will need to be completed by the end of December, otherwise it will need to be postponed to the 2023.1 release.

**Comment by josephsl** [comment]
Hi, at this point, I think the best strategy is adopting IAccessible handler reorg – convert UIA handler into a package with submodules representing constants, event handlers, and other goodies. If I do come around to doing the reorg, the latest time I can work on it will be early December. Thanks for the reminder.

**Comment by josephsl** [comment]
Hi,

Indeed, I'm doing the reorg at the moment (changed the issue title accordingly) as follows:

1. Replace _UIAHandler (private module) usage with UIAHandler (public module). So far Excel UIA object definitions module is using this approach.
2. Create "UIAHandler" package with all UIA modules (public/private main module, constatns, browse mode, utilities, etc.) moved into it.
3. Constants and other modules will become submodules of UIA handler package and should be used as UIAHandler.constants, for example.

I expect the initial work (and a pull request) will be done by American Thanksgiving Day (November 25, 2021) so it can be reviewed for 2022.1.

Thanks.

**Comment by Brian1Gaff** [comment]
Hi did this go out to individuals instead of one of the mailing lists?
 Brian

***@***.***
Sent via blueyonder.
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users
----- Original Message -----
